### PR TITLE
Add post to subreddit experiment

### PR DESCRIPTION
**Pull request by keithblaha**
:eyeglasses: @scarow @schwers 

@schwers - i heard the experiment framework for 2X was/is being rewritten, am i using it right?

note: requires https://github.com/reddit/node-api-client/pull/179

### CODE REVIEW

File: src/app/reducers/subredditsByPost.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import * as subredditsByPostActions from 'app/actions/subredditsByPost';
+
+const DEFAULT = {};
+
+export default (state=DEFAULT, action={}) => {
+  switch (action.type) {
+    case subredditsByPostActions.RECEIVED_SUBREDDITS_BY_POST: {
+
+      return merge(state, {
+        [action.postId]: Object.values(action.apiResponse.subreddits),
```
Comment (schwers) [inline src/app/reducers/subredditsByPost.js]: We should track the subreddits by id here, and merge the models in reducers/subreddits

File: src/app/featureFlags.js [hunk 2]
```diff
@@ -74,6 +75,14 @@ const config = {
       seoReferrer: true,
     }],
   },
+  [VARIANT_RECOMMENDED_BY_POST]: {
+    url: 'experimentrecommendedbypostsrbypost',
+    and: [{
+      variant: 'subreddits_by_post:sr_by_post',
```
Comment (schwers) [inline src/app/featureFlags.js]: The logged in check should be first so we don't fire bucketing events for users that are inelgible

**Comment by keithblaha** [comment]
:haircut_woman:

**Comment by keithblaha** [comment]
bump and adding :eyeglasses: @nramadas

**Comment by keithblaha** [comment]
FYI i'll be adding 1 more thing (some click event tracking)

File: src/app/actions/subredditsToPostsByPost.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    if (some(LEGAL_VARIANTS, variant => feature.enabled(variant))) {
+      dispatch(fetching(post.id));
+      const apiOptions = apiOptionsFromState(state);
+
+      try {
+        const response = await SubredditsToPostsByPost.get(
+          apiOptions,
+          {
+            link: post.name,
+            variant: 'nb',
```
Comment (scarow) [inline src/app/actions/subredditsToPostsByPost.js]: What is 'nb'?

File: src/app/components/RecommendedSubreddits/index.jsx [hunk 4]
```diff
@@ -37,7 +37,7 @@ export default function RecommendedSubreddits(props) {
     subredditListing = (
       <div className={ cssClass }>
         <Anchor
-          href={ addUtmTracking(currentSubreddit.url, 0, variant) }
+          href={ currentSubreddit.url }
```
Comment (scarow) [inline src/app/components/RecommendedSubreddits/index.jsx]: Data team doesn't want utm tracking?

File: src/app/pages/Comments.jsx [hunk 5]
```diff
@@ -98,6 +119,8 @@ CommentsPage.propTypes = {
   currentPage: T.object.isRequired,
   preferences: T.object.isRequired,
   isCrawlerRequest: T.bool.isRequired,
+  recommendedSubreddits: T.arrayOf(T.object).isRequired,
+  recommendedPosts: T.arrayOf(T.object).isRequired,
```
Comment (scarow) [inline src/app/pages/Comments.jsx]: Should these be required for all `CommentsPage`? Not sure

File: src/app/pages/Comments.jsx [hunk 6]
```diff
@@ -121,8 +144,30 @@ const stateProps = createSelector(
...
+  },
+  (state, props) => {
+    const pageParams = CommentsPageHandler.pageParamsToCommentsPageParams(props);
+    const post = state.posts[pageParams.id];
+    const postLoaded = !!post;
+    let recommendedPosts = null;
+    if (postLoaded && post.id in state.subredditsToPostsByPost) {
+      const postNames = state.subredditsToPostsByPost[post.id];
+      recommendedPosts = Object.values(state.posts).filter(post => postNames.includes(post.uuid));
+    }
```
Comment (scarow) [inline src/app/pages/Comments.jsx]: Can `postNames` ever be `null`? If so, what happens? Does this break?

File: src/app/pages/Comments.jsx [hunk 7]
```diff
@@ -133,6 +178,8 @@ const stateProps = createSelector(
       currentPage,
       preferences,
       isCrawlerRequest,
+      recommendedSubreddits: !recommendedSubreddits ? [] : recommendedSubreddits,
+      recommendedPosts: !recommendedPosts ? [] : recommendedPosts,
```
Comment (scarow) [inline src/app/pages/Comments.jsx]: IMO it's easier to read this without the `!`, thereby flipping the logic. Like:

`recommendedPosts: recommendedPosts ? recommendedPosts : [],`

Up to you

File: src/app/pages/Comments.jsx (hunk 6 above)
Comment (keithblaha) [inline src/app/pages/Comments.jsx]: the check inside if statement should prevent that from happening

File: src/app/pages/Comments.jsx (hunk 7 above)
Comment (keithblaha) [inline src/app/pages/Comments.jsx]: i agree but i was following the general style throughout the file (better to be locally correct than globally correct)

**Comment by keithblaha** [comment]
bump- please review, tracking is now in and this is functionally complete. we are trying to deploy this tomorrow

**Comment by keithblaha** [comment]
adding :eyeglasses: @prashtx @nramadas

File: src/app/pages/Comments.jsx (hunk 5 above)
Comment (keithblaha) [inline src/app/pages/Comments.jsx]: it's expected that itll be [] if none

File: src/app/components/RecommendedSubreddits/styles.less [hunk 8]
```diff
@@ -68,10 +73,6 @@
     margin: 0px 40px;
     padding: 10px 10px 0px 10px;
     text-align: center;
-
-    .themeify({
-      color: @theme-toast-color;
-    });
```
Comment (scarow) [inline src/app/components/RecommendedSubreddits/styles.less]: Can;t remember what this is for -- can you attach a screen shot of what this looks like in night mode?

File: src/app/router/handlers/CommentsPage.js [hunk 9]
```diff
@@ -88,8 +94,9 @@ export default class CommentsPage extends BaseHandler {
       subredditName = post.subreddit;
     }
 
-    if (subredditName) {
-      fetchRecommendedSubreddits(state, dispatch, subredditName);
+    if (post) {
+      fetchRecommendedSubredditsByPost(state, dispatch, post);
+      fetchRecommendedSubredditsToPostsByPost(state, dispatch, post);
```
Comment (scarow) [inline src/app/router/handlers/CommentsPage.js]: Why are we fetching both of these? Isn't this dependent on variant?

File: src/app/components/RecommendedSubreddits/styles.less (hunk 8 above)
Comment (keithblaha) [inline src/app/components/RecommendedSubreddits/styles.less]: been through review with design team and they're happy with how it looks:

subreddits_nightv5

File: src/app/router/handlers/CommentsPage.js (hunk 9 above)
Comment (keithblaha) [inline src/app/router/handlers/CommentsPage.js]: the bucketing check is inside the fetch functions

File: src/lib/eventUtils.js [hunk 10]
```diff
@@ -85,6 +85,17 @@ function trackScreenViewEvent(state, additionalEventData) {
...
 
+export function trackExperimentClickEvent(state, experimentName, targetThing) {
+  const payload = {
+    ...getBasePayload(state),
+    'experiment_name': experimentName,
+    'target_fullname': targetThing.name,
+    'target_url': targetThing.url,
+    'target_type': targetThing.type,
+  };
+  getEventTracker().track('internal_click_events', 'cs.experiment_clicks', payload);
```
Comment (scarow) [inline src/lib/eventUtils.js]: has this been approved by data? Not sure what event names should look like.

File: src/app/router/handlers/CommentsPage.js [hunk 11]
```diff
@@ -116,6 +123,35 @@ const fetchRecommendedSubreddits = (state, dispatch, subredditName) => {
...
 };
 
+const fetchRecommendedSubredditsByPost = (state, dispatch, post) => {
+  let experimentId = 105;
+  dispatch(subredditsByPostActions.fetchSubredditsByPost(post, experimentId));
+};
+
+const fetchRecommendedSubredditsToPostsByPost = (state, dispatch, post) => {
+  const feature = features.withContext({ state });
+  let experimentId = 105;
```
Comment (scarow) [inline src/app/router/handlers/CommentsPage.js]: Should we store `experimentId` so we don't have to dupe it in `fetchRecommendedSubredditsByPost` and `fetchRecommendedSubredditsToPostsByPost`?

File: src/lib/eventUtils.js (hunk 10 above)
Comment (keithblaha) [inline src/lib/eventUtils.js]: it's already speced out https://reddit.atlassian.net/wiki/display/DA/Internal+Click+Events

File: src/app/router/handlers/CommentsPage.js (hunk 11 above)
Comment (keithblaha) [inline src/app/router/handlers/CommentsPage.js]: ultimately thisll be ripped out when its featurized so probably not worth the effort

File: src/app/components/Post/index.jsx [hunk 12]
```diff
@@ -182,7 +185,7 @@ export function Post(props) {
           showingLink={ !!(displayCompact && !hasExpandedCompact && externalDomain) }
           renderMediaFullbleed={ renderMediaFullbleed }
           showLinksInNewTab={ showLinksInNewTab }
-          onElementClick={ onElementClick }
+          onElementClick={ () => { onPostClick(post); onElementClick() } }
```
Comment (keithblaha) [inline src/app/components/Post/index.jsx]: there should be full internal click tracking across the board.... but, note that this is an experiment click event, not a generic click event. so we should only fire these types of events for experiments

**Comment by scarow** [comment]
🐟  lgtm -- I would wait on the review of an mweb person for any subtleties I might be missing!

**Comment by keithblaha** [comment]
thank you @scarow!

we are about a week late pushing this experiment out, so anyone who'd like to review it please try to today! aiming to deploy this tomorrow

**Comment by prashtx** [comment]
It looks like there are some linter errors that travis-ci is catching.

File: src/app/reducers/subreddits.js [hunk 13]
```diff
@@ -17,6 +18,7 @@ export default function(state=DEFAULT, action={}) {
 
     case recommendedSubredditsActions.RECEIVED_RECOMMENDED_SUBREDDITS:
     case searchActions.RECEIVED_SEARCH_REQUEST:
+    case subredditsByPostActions.RECEIVED_SUBREDDITS_BY_POST:
```
Comment (prashtx) [inline src/app/reducers/subreddits.js]: Can you add this to the tests for this reducer?

**Comment by prashtx** [comment]
💅

File: src/app/reducers/subreddits.js (hunk 13 above)
Comment (keithblaha) [inline src/app/reducers/subreddits.js]: is that necessary? it is tested in its own file (included in this PR)

File: src/app/reducers/posts.js [hunk 14]
```diff
@@ -45,6 +46,7 @@ export default function(state=DEFAULT, action={}) {
     case hiddenActions.RECEIVED_HIDDEN:
     case savedActions.RECEIVED_SAVED:
     case searchActions.RECEIVED_SEARCH_REQUEST:
+    case subredditsToPostsByPostActions.RECEIVED_SUBREDDITS_TO_POSTS_BY_POST:
```
Comment (keithblaha) [inline src/app/reducers/posts.js]: is that necessary? it is tested in its own file (included in this PR)

**Comment by keithblaha** [comment]
:haircut_woman:

File: src/app/pages/Comments.jsx [hunk 15]
```diff
@@ -121,8 +152,30 @@ const stateProps = createSelector(
...
   state => state.platform.currentPage,
   state => state.preferences,
+  (state, props) => {
+    const pageParams = CommentsPageHandler.pageParamsToCommentsPageParams(props);
+    const post = state.posts[pageParams.id];
+    const postLoaded = !!post;
+    let recommendedSubreddits = null;
+    if (postLoaded && post.id in state.subredditsByPost) {
+      const subredditNames = state.subredditsByPost[post.id];
+      recommendedSubreddits = Object.values(state.subreddits).filter(subreddit => subredditNames.includes(subreddit.uuid));
```
Comment (schwers) [inline src/app/pages/Comments.jsx]: I want to do a quick test, but I think using `Object.values` like this will end up creating a new object everytime state changes, which could cause re-renders for any state change. The first pass I usually do for testing this page is to log in and go to  heavily nested comments page (usually anything on the logged out frontpage or askreddit), go to timeline in dev tools and turn on 5x device throttling, start measuring, and upvote a nested comment and then stop measuring. It usually takes somewhere between 100ms and 200ms to handle the click event in prod with redux dev tools enabled.

File: src/app/reducers/subreddits.js (hunk 13 above)
Comment (schwers) [inline src/app/reducers/subreddits.js]: +1 for test

File: src/app/pages/Comments.jsx (hunk 15 above)
Comment (prashtx) [inline src/app/pages/Comments.jsx]: `filter` will create a new object either way. If this causes unnecessary rerenders and a related performance hit, would it be better to move the filtering logic into the combiner, so that the selectors produce the exact same objects for unrelated state changes?

Comment (schwers) [inline src/app/pages/Comments.jsx]: Good point, and from testing `values` does a create a new object as well. 

From my tests there was a performance hit to voting and comment collapse, I'd suspect there's more than those two interactions given how this is tied to any state change. Ideally we should only be returning new object references from a selector like this if the data we need has changed. In this case I think we should be able to return `state.subredditsByPost[post.id]` to avoid new object references. At that point we'd either have to pass `state.subreddits` into the combiner to extract the subreddit objects, or connect the `RecommendedSubreddits` component to the store and pass it the list of recommended subreddit ids that it would extract in its own selector.

In the combiner approach: `state.subreddits` shouldn't change that often, so this would minimize the re-renders. The next selector for `subredditsToPostsByPost` would need to do something similar, but would require passing `state.posts` into the combiner as well. Given that would make for two 'hacks', I think it would be cleanest to connect the `RecommendedPosts` and `RecommendedSubreddits` components and pass them a list of id's they should be rendering. For this selector that would mean returning `state.subredditsByPost[post.id]` and in the next selector, `state.subredditsToPostsByPost[post.id]`.

**Comment by schwers** [comment]
💅

File: src/app/components/RecommendedPosts/styles.less [hunk 16]
```diff
@@ -0,0 +1,72 @@
+@import (reference) '~app/less/variables';
+@import (reference) '~app/less/themes/themeify';
+
+.RecommendedPosts__container {
+  .themeify({
+    background-color: @theme-comment-body-color;
+  });
+  .nightMode & {
```
Comment (keithblaha) [inline src/app/components/RecommendedPosts/styles.less]: well thats handy!

File: src/app/pages/Comments.jsx (hunk 15 above)
Comment (keithblaha) [inline src/app/pages/Comments.jsx]: thank you for investigating, definitely wanna minimize performance impact. ill refactor as suggested

**Comment by keithblaha** [comment]
partial :haircut_woman: 

i still need to add those reducer tests (will do tomorrow), but i did the refactoring for the recommendation components to be connected which i think will be the harder part to review

**Comment by keithblaha** [comment]
added reducer tests. FYI i'll be adding one more endpoint to this to support another variant in the experiment... a direct post to post recommendation

File: src/app/pages/Comments.jsx [hunk 17]
```diff
@@ -51,6 +53,14 @@ function CommentsPage(props) {
     <div className='CommentsPage'>
       <SubNav />
       <Post postId={ pageParams.id } single={ true } key='post' />
+      <RecommendedPosts
+        postId={ pageParams.id.substring(3) }
```
Comment (schwers) [inline src/app/pages/Comments.jsx]: This is a really nice refactor, thanks for doing that! Just a small nit but I have a pretty strong preference for using the full post id so we're as opaque about ids as possible.

**Comment by schwers** [comment]
Just as general note, @keithblah, @curioussavage is taking over a CONT ticket PR, https://github.com/reddit/reddit-mobile/pull/920, and these do sort of similiar things. Can you coordinate and make sure we don't show two of these relevancy experiments on the same comments page?

File: src/app/pages/Comments.jsx (hunk 17 above)
Comment (keithblaha) [inline src/app/pages/Comments.jsx]: Good point- and I agree, it's my bad for using ids instead of fullnames in the actions!

**Comment by keithblaha** [comment]
Thanks for letting me know, I'll talk with @curioussavage about it tomorrow

This PR is finally in a final state now- no more functionality to add. We're aiming to push live monday! I'll bug you all for fish tomorrow :)

**Comment by keithblaha** [comment]
Friendly bump- would love to have this fished today so we can launch on monday!

File: src/app/components/RecommendedPosts/styles.less [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    padding: 0;
+    width: 90%;
+
+    .themeify({
+      border-top: @theme-border;
+    });
+  }
+
+  .Post__header-wrapper {
+    padding-left: 16px;
```
Comment (schwers) [inline src/app/components/RecommendedPosts/styles.less]: small nit: I would try to use a multiple of `@grid-size` whenever possible for alignment and spacing like this. It will be more consistent across the other less files in the repo and keep things uniform visually

File: src/app/components/RecommendedSubreddits/index.jsx [hunk 19]
```diff
@@ -1,104 +1,113 @@
...
-                    : {}
-                  }
-          />
+  const subredditListing = recommendedSubreddits.map((sr) => {
+    return (
+      <div className='RecommendedSubreddits__top'>
+        <div
+          className='subreddit-icon-image'
+          style={ sr.iconImage
+                  ? { 'backgroundImage': `url(${sr.iconImage})`,
```
Comment (schwers) [inline src/app/components/RecommendedSubreddits/index.jsx]: Sorry for missing this, but you should use the keyColor here as well. There are some subreddits that put transparency in their icon and rely on the keycolor showing through for branding / themeing

File: src/app/components/RecommendedSubreddits/styles.less [hunk 20]
```diff
@@ -23,6 +22,10 @@
   }
 }
 
+.RecommendedSubreddits__SubredditInfo {
+  margin-top: 8px;
```
Comment (schwers) [inline src/app/components/RecommendedSubreddits/styles.less]: Looks like a good place to use `@grid-size` too

File: src/app/components/RecommendedSubreddits/styles.less [hunk 21]
```diff
@@ -34,19 +37,20 @@
...
+   }, @inNightMode: {
+      background-color: #222;
   });
 
   .subreddit-icon-image {
     width: @community-icon-size;
     height: @community-icon-size;
     float: left;
-    margin-right: 25px;
+    margin-right: 16px;
```
Comment (schwers) [inline src/app/components/RecommendedSubreddits/styles.less]: Another use case for `@grid-size`

File: src/app/components/RecommendedPosts/styles.less (hunk 18 above)
Comment (keithblaha) [inline src/app/components/RecommendedPosts/styles.less]: so i mostly based the css off the old PR for 1X for similar subreddits experiment... sorry its not up to snuff. the good news is that design has been working on new designs for recommendations, so all of this CSS is throwaway!

File: src/app/router/handlers/CommentsPage.js [hunk 22]
```diff
@@ -6,16 +6,23 @@ import { cleanObject } from 'lib/cleanObject';
 import { fetchUserBasedData } from './handlerCommon';
 import * as commentsPageActions from 'app/actions/commentsPage';
 import * as subredditActions from 'app/actions/subreddits';
-import * as recommendedSubredditsActions from 'app/actions/recommendedSubreddits';
+//import * as recommendedSubredditsActions from 'app/actions/recommendedSubreddits';
```
Comment (keithblaha) [inline src/app/router/handlers/CommentsPage.js]: so were still working on the similar subreddits model (which goes subreddit->subreddit, as opposed to post->subreddit which were testing this PR)... so this might be something we use again. just wanted to silence lint errors as its not used

File: src/app/components/RecommendedSubreddits/index.jsx (hunk 19 above)
Comment (keithblaha) [inline src/app/components/RecommendedSubreddits/index.jsx]: good call, will do
